libraries/AP_RCProtocol/AP_RCProtocol_MavlinkRadio.cpp

```diff
@@ -0,0 +1,31 @@
+
+#include "AP_RCProtocol_config.h"
+
+#if AP_RCPROTOCOL_MAVLINK_RADIO_ENABLED
+
+#include "AP_RCProtocol_MAVLinkRadio.h"
+
+// constructor
+AP_RCProtocol_MAVLinkRadio::AP_RCProtocol_MAVLinkRadio(AP_RCProtocol &_frontend) :
+    AP_RCProtocol_Backend(_frontend)
+{}
+
+void AP_RCProtocol_MAVLinkRadio::update_radio_rc_channels(const mavlink_radio_rc_channels_t* packet)
+{
+    uint8_t count = packet->count;
+    if (count >= MAX_RCIN_CHANNELS) count = MAX_RCIN_CHANNELS;
+
+    uint16_t rc_chan[MAX_RCIN_CHANNELS];
+    for (uint8_t i = 0; i < count; i++) {
+        // The channel values are in centered 13 bit format. Range is [-4096,4096], center is 0.
+        // According to specification, the conversion to PWM is x * 5/32 + 1500.
+        rc_chan[i] = ((int32_t)packet->channels[i] * 5) / 32 + 1500;
+    }
+
+    bool failsafe = (packet->flags & RADIO_RC_CHANNELS_FLAGS_FAILSAFE);
```

Comment:
This is not sufficient.  If the values are frozen we don't want to use them either!

Reply:
I'm not sure what you mean by "froozen"
the case when frames are not received is handled by the upper level code
if you mean handling missing frames bit: IMHO this is exactly as for sbus